### ISSUE TITLE
change `Codec::encode` to take `self` by value

### ISSUE THREAD

**Issue by guswynn** [MaterializeInc/materialize#12341]
I noticed in https://github.com/MaterializeInc/materialize/pull/12297 that `encode` takes `&self`, which means that the `From<ProtoVersion>` impls around the codebase actually are implemented as `From<&ProtoVersion>`, and internally call `.clone()` on sub-fields.

This forces users who have already owned versions of structs to call `clone` unnecessarily. I propose we change `Codec` to:

```
pub trait Codec: Sized + 'static {
    fn encode<B>(self, buf: &mut B)
    where
    ¦   B: BufMut;

    fn encode_ref<B>(&self, buf: &mut B)
    where
    ¦   B: BufMut, Self: Clone {
        self.clone().encode(buf)
    }

    fn decode<'a>(buf: &'a [u8]) -> Result<Self, String>;
}
```
(we could also lift the `Clone` bound to a supertrait)

In addition to this change, we can also switch the `From` impls to `From<OwnedProtoVersionOfStruct> for Struct` as opposed to by reference.

This means that owned structs can just use `encode`, but places where we have only shared access can call `encode_ref` _which will clone the exact same number of things as we do now_

**Comment by danhhz** [comment]
I don't think this is quite right. Most actual encodings don't get anything from taking an owned `self`: they end up memcpy'ing anyway. IMO `encode` should be able to take `&self` and still not need a clone and this is really an artifact of prost's default generated code. Concretely:

we have some rust type Foo:
```
struct Foo{
    pub bar: String,
}
```

to encode it via prost:
```
struct ProtoFoo {
    pub bar: String,
}
```

we indeed either need to take an owned Foo or clone the string. But the actual proto encoding doesn't need bar to be owned, it just ends up getting memcpy'd. If we instead has something like

```
struct ProtoFoo<'a> {
    pub bar: &'a String,
}
```

we could have our cake and eat it too. I wonder if https://github.com/tokio-rs/prost/pull/341 might help with this?

**Comment by guswynn** [comment]
proto structs that take references would certainly be nice, but I doubt that will ever happen; I think it gets really hairy when deciding what do with sub-messages. What would the `Bytes` help with other than `Vec<u8>` fields?

**Comment by guswynn** [comment]
I think we should still change the from impls, but leave `encode` as taking `&self`, but the discussion has continued here: https://github.com/MaterializeInc/materialize/issues/12579

**Comment by aalexandrov** [comment]
For completeness, part of the discussion was in #12579. I don't think we can address as part of #12579, so it makes sense to continue here.

**Comment by aalexandrov** [comment]
@guswynn let's use this issue to iron out the details in #12579.

**Comment by guswynn** [comment]
@aalexandrov I think at this point, we should
- Leave `Codec::encode` as taking `&self` (cc @danhhz)
- Switch `into_proto` to take `RustType` _by value_ not `&self`, then, the `Codec::encode` impl for `RustType` (do we implement `Codec` for `RustType`s?) can be:
```
fn encode(&self, &mut B) {
    self.encode(self.clone().into_proto())`
}
```

This should be the same number of `clones` (cc @petrosagg), as implementors of `into_proto` will do the same thing: basically call `into_proto` on the subfields, but those will also take the field by value, and for things like `Strings`, we will just _move_ the string instead of clone it

This improves clarity and consistency with `from_proto`, and is similar to how `From/Into` in `std` work. Additionally, callsites that have owned access to a `RustType` can do `rust_type.into_proto().encode()` to just do moves, not clones before encoding
